README.md
# waterline-sails.io.js Adapter

This module is a [Waterline](https://github.com/balderdashy/waterline)/[Sails.js](https://github.com/balderdashy/sails) 
adapter which provides easy access to a Sails server via the 
[sails.io.js web-socket client](http://sailsjs.org/documentation/reference/web-sockets/socket-client), when
Waterline is used on a client. It is designed to work with 
[Sails Blueprint](http://sailsjs.org/documentation/reference/blueprint-api) routes so it can be used to hook up
a client application to a Sails.js server right out the box. 

This adapter is particularly useful when creating isomorphic applications where records must be retrieved on the
client or server in a consistent manner. On the server the models can be configured to use the adapters of the 
underlying persistence layer/services, and when passing the model config to the client the connection details
can be updated to use this `sails.io.js` adapter instead.
 

### Installation

Install via NPM:

```sh
$ npm install waterline-sails.io.js
```


### Configure

Add the sails.io.js adapter config to the config/connections.js file. Available options and defaults:

```javascript
module.exports.connections = {
  sailsSocketClient: {
    adapter: 'waterline-sails.io.js',
    protocol: 'http',   // HTTP protocol (http | https)
    host: 'localhost',  // api host name
    port: 1337,         // api port
    basePath: '/api',   // base api path
    autoConnect: true,  // eager/lazy connect the web-socket
    simulateDelay: 0    // Simulated delay ms
  }
};
```

You will have to update the `connection` property on the definitions of the models that will be used
on the client so it uses the `sailsSocketConnection` rather than the connection specified on the server.

See the documentation of 
[fluxible-plugin-waterline-models](https://github.com/marnusw/fluxible-plugin-waterline-models#use-with-sailsjs)
for an example of how this might be configured.

##### Request URLs

The request URLs match the [Sails Blueprints](http://sailsjs.org/documentation/reference/blueprint-api) specification;
that is: 

`REST_METHOD <protocol>://<host>:<port><basePath>/<model|collectionName>[/:id]`

There are two exceptions to this rule; see the [usage](#usage) section.

##### Simulated Response Delay

Since the adapter will be used on a client it may at times be convenient to simulate the delay of a 
server round trip while developing on a local server. By specifying the `simulateDelay` option in 
milliseconds on the config object all requests will be delayed by that amount of time. This feature 
is automatically disabled and removed from a production build.


### Interfaces and Features

The following 
[**Interfaces**](https://github.com/balderdashy/sails-docs/blob/master/contributing/adapter-specification.md) are 
supported:

* Semantic
* Queryable

The `.populate()` method without criteria, i.e. calling `Model.populate('attrName')`, is allowed, but with
`Model.populate('attrName', {criteria})` the criteria will be ignored. All attributes passed on successive
calls to `.populate()` will be passed to the server on the `populate` property of the query. This means
that if [Sails Blueprints](http://sailsjs.org/documentation/reference/blueprint-api) are used this will 
work straight out of the box.

The supported **Features** will depend on the underlying storage mechanism used on the server, but in 
principle the following features may be supported:

* `autoIncrement`(`.sequential`)
* `unique`

**Planned Features**

* `crossAdapter` (queries across this and other adapters will be joined in-memory on the client)

At present this is not implemented and only joins via a single connection using this adapter is supported.


### Usage

This adapter exposes the standard `find()`, `create()`, `update()` and `destroy()` methods and other 
combinations of the *Semantic Interface* and can be used on the client just like any of the core adapters 
would be on the server, making this adapter an attractive choice for isomorphic applications. 
Sails Blueprints does not support the `findAndUpdate()` or `findAndDestroy()` methods, but this 
[semantic-blueprints hook](https://github.com/marnusw/sails-hook-semantic-blueprints) can be used to 
extend your app with those.

As always the `.exec()` method will return an error object or the results, or the promise api can be used.

##### `sailsIoSocket()` Method

The adapter also exposes a `sailsIoSocket()` method on the Model which serves the same purpose as the 
`.native()` and `.query()` methods on the MongoDB and SQL adapters respectively. It can be used to run 
one of the `sails.io.js` methods against the actual underlying socket. 

```javascript
Model.sailsIoSocket(method, url, data, function(err, body, jwres) {...});

// OR

Model.sailsIoSocket(method, url, data).then(body => {...}).catch(err => {...});
```

The first parameter is the `sails.io.js` method to call, with subsequent parameters being passed as the 
arguments. The data parameter is optional, and the callback should always be the last argument provided.
If the callback is omitted a promise will be returned. In this case it is not possible to gain access to 
the raw `jwres` object.

*Note:* When a callback is not provided a global `Promise` class is assumed, meaning this case must either 
be in an environment which supports ES6 features or be polyfilled. Including the 
[Babel polyfill](https://babeljs.io/docs/usage/polyfill/), 
[es6-promise](https://www.npmjs.com/package/es6-promise) or any other polyfill with the ES6 API will work.
If callbacks are used throughout it is not necessary to add a promise polyfill.

The only difference compared to using the socket methods directly is the signature of the callback. The 
first parameter is a possible error object (see below), the response body is on the second parameter 
for successful requests, and the untouched `jwres` object is always provided on the third parameter for 
more specialised use cases. 

##### `addTo()` and `removeFrom()` Methods

The adapter also exposes `addTo()` and `removeFrom()` methods on the Model for gaining access to the
Blueprints [Add to Collection](http://sailsjs.org/documentation/reference/blueprint-api/add-to) and
[Remove from Collection](http://sailsjs.org/documentation/reference/blueprint-api/remove-from) actions.

```javascript
var options = {
  id: 1,                    // The primary record to update.
  association: 'attrName',  // The association attribute.
  foreignId: 3,             // The id of an existing record to add onto/remove from the association, OR
  foreignData: {...}        // Data to create a new record to add onto/remove from the association.
};

// Then do either

Model.addTo(options, function(err, primaryRecordData, jwres) {...});
Model.addTo(options).then(primaryRecordData => {...}).catch(err => {...});

// or

Model.removeFrom(options, function(err, primaryRecordData, jwres) {...});
Model.removeFrom(options).then(primaryRecordData => {...}).catch(err => {...});
```

In both cases the `primaryRecordData` is the data of the record added to, with all records on the
relevant association populated. Because these are not native Sails.js methods, however, this is
only the record data and not a Waterline record instance.

*Note:* When a callback is not provided a global `Promise` class is assumed, meaning this case must either 
be in an environment which supports ES6 features or be polyfilled. Including the 
[Babel polyfill](https://babeljs.io/docs/usage/polyfill/), 
[es6-promise](https://www.npmjs.com/package/es6-promise) or any other polyfill with the ES6 API will work.
If callbacks are used throughout it is not necessary to add a promise polyfill.

##### Error Objects

If a web socket request responds with an error an error-object is returned as the first argument of
the callback. This will be a `WLError` instance created by Waterline, the error objects below will 
be on the `originalError` property. The exception is native calls to `sailsIoSocket()`, where these
objects are returned directly.

For a 400 status code (validation errors) the object has the form:

```javascript
  {
    statusCode: 400,
    type: 'E_VALIDATION',
    validation: true,
    messages: jwres.body
  }
```

For a 401 status code (unauthorized) the object has the form:

```javascript
  {
    statusCode: 401,
    type: 'E_UNAUTHORIZED',
    unauthorized: true,
    validation: true,
    messages: jwres.body
  }
```

For a 404 status code (not found) the object has the form:

```javascript
  {
    statusCode: 404,
    type: 'E_NOT_FOUND',
    notFound: true,
    messages: jwres.body
  }
```

For all other status codes the object has the form:

```javascript
  {
    statusCode: <code>,
    type: 'E_FATAL',
    fatal: true,
    message: jwres.body
  }
```


### Tests

No tests yet. The [Waterline-adapter-tests](https://github.com/balderdashy/waterline-adapter-tests) should be run.


## Contributing

Comments and/or pull requests are welcome.


## License

This software is free to use under the MIT license.
See the [LICENSE file](LICENSE.md) for license text and copyright information.


CHANGELOG.md
### 0.5.1 (29 July 2015)

* [FEATURE] Direct access to sails.io.js via `Model.sailsIoSocket()` supports the promise api.
* [FEATURE] Added `addTo()` and `removeFrom()` methods for gaining access to the Blueprints add to/remove from actions.
* [ENHANCEMENT] Direct access to sails.io.js via `Model.sailsIoSocket()` forces the `method` to lower case.
* [BUGFIX] Direct access to sails.io.js via `Model.sailsIoSocket()` honours `simulateDelay`.

### 0.5.0 (6 July 2015)

* [DOCUMENTATION] Include a note on WLErrors wrapping the error objects.
* [ENHANCEMENT] Removed the dependency on `waterline-errors`.
* [ENHANCEMENT] Cleaned up the codebase significantly.
* [FEATURE] Renamed the `pathname` config option to `basePath`.
* [FEATURE] Added the `autoConnect` config option rather than forcing eager connection.
* [DOCUMENTATION] Fix: `crossAdapter` feature is planned, not implemented.
* [DOCUMENTATION] Document the `simulateDelay` config option.

### 0.4.0 (4 July 2015)

* [DOCUMENTATION] Wrote up the main README.md documentation. 
* [ENHANCEMENT] Removed the dependency on `lodash`. 
* [FEATURE] Allow configuring a simulated delay (with `simulateDelay` option) in non-production environments. 
* [FEATURE] Supporting `.populate` on queries via the `.join()` method which passes it on the query. 

### 0.3.1 (13 June 2015)

* [FEATURE] Pass back proper E_NOT_FOUND error on 404's rather than a fatal error. 

### 0.3.0 (2 June 2015)

* [FEATURE] Direct access to sails.io.js methods via the `Model.sailsIoSocket()` method.
* [FEATURE] Pass back proper E_UNAUTHORIZED error on 401's rather than a fatal error. 

### 0.2.0 (25 May 2015)

* [FEATURE] Pass back an E_VALIDATION error 400 errors and E_FATAL errors in all other cases
 rather than `jwres.body`. The latter is passed as the `message` property of error object. 

### 0.1.1 (25 May 2015)

Initial release.


test/unit/register.js
/**
 * Test dependencies
 */
var Adapter = require('../../lib/adapter_old');


describe('registerCollection', function () {

	it('should not hang or encounter any errors', function (done) {
		Adapter.registerCollection({
			identity: 'foo'
		}, done);
	});

	// e.g.
	// it('should create a mysql connection pool', function () {})
	// it('should create an HTTP connection pool', function () {})
	// ... and so on.
});


test/integration/runner.js
/**
 * Test runner dependencies
 */
var util = require('util');
var mocha = require('mocha');
var log = new (require('captains-log'))();


var TestRunner = require('waterline-adapter-tests');
var Adapter = require('../../lib/adapter');


// Grab targeted interfaces from this adapter's `package.json` file:
var packageJson = {};
var interfaces = [];
try {
  packageJson = require('root-require')('package.json');
  interfaces = packageJson['waterlineAdapter'].interfaces;
} catch (e) {
  throw new Error(
    '\n' +
    'Could not read supported interfaces from "waterlineAdapter"."interfaces"' + '\n' +
    'in this adapter\'s `package.json` file ::' + '\n' +
    util.inspect(e)
  );
}


log.info('Testing `' + packageJson.name + '`, a Sails adapter.');
log.info('Running `waterline-adapter-tests` against ' + interfaces.length + ' interfaces...');
log.info('( ' + interfaces.join(', ') + ' )');
console.log();
log('Latest draft of Waterline adapter interface spec:');
log('https://github.com/balderdashy/sails-docs/blob/master/adapter-specification.md');
console.log();


/**
 * Integration Test Runner
 *
 * Uses the `waterline-adapter-tests` module to
 * run mocha tests against the specified interfaces
 * of the currently-implemented Waterline adapter API.
 */
new TestRunner({

  // Load the adapter module.
  adapter: Adapter,

  // Default adapter config to use.
  config: {
    schema: false
  },

  // The set of adapter interfaces to test against.
  // (grabbed these from this adapter's package.json file above)
  interfaces: interfaces

  // Most databases implement 'semantic' and 'queryable'.
  //
  // As of Sails/Waterline v0.10, the 'associations' interface
  // is also available.  If you don't implement 'associations',
  // it will be polyfilled for you by Waterline core.  The core
  // implementation will always be used for cross-adapter / cross-connection
  // joins.
  //
  // In future versions of Sails/Waterline, 'queryable' may be also
  // be polyfilled by core.
  //
  // These polyfilled implementations can usually be further optimized at the
  // adapter level, since most databases provide optimizations for internal
  // operations.
  //
  // Full interface reference:
  // https://github.com/balderdashy/sails-docs/blob/master/adapter-specification.md
});


lib/socket.js
/**
 * Copyright 2015, Marnus Weststrate
 * Copyrights licensed under the MIT License. See the accompanying LICENSE file for terms.
 */
var socketIOClient = require('socket.io-client');
var sailsIOClient = require('sails.io.js');


/**
 * A sails.io.js socket connection.
 *
 * @constructor
 * @param {Object} config
 * @param {Object} collections
 */
var Socket = module.exports = function(config, collections) {

  // Instantiate the socket client `io` (for now the socket.io client must be passed in explicitly).
  this.io = sailsIOClient(socketIOClient);

  this.io.sails.autoConnect = config.autoConnect;
  this.io.sails.url = config.protocol + '://' + config.host + ':' + config.port;

  this.basePath = (config.basePath || '') + '/';

  this.collections = collections || {};
  this.config = config;

  return this;
};

// Save a couple of bytes after minification
var SocketPrototype = Socket.prototype;

/**
 * Close the socket connection.
 */
SocketPrototype.teardown = function() {
  this.io.socket.disconnect();
};


///////////////////////////////////////////////////////////////////////////////////////////
/// DIRECT SOCKET ACCESS
///////////////////////////////////////////////////////////////////////////////////////////

/**
 * Run one of the sails.io.js methods against the actual socket. The callback signature
 * differs in that it is a Node like callback with a possible error object returned on
 * the first parameter, the response body on the second parameter for successful requests,
 * and the untouched `jwres` object on the third parameter for special use cases:
 * `function(err, body, jwres)`
 *
 * The first parameter is the method to call, with subsequent parameters being passed as
 * the arguments. The callback should always be the last argument provided.
 *
 * @param {String} method
 */
SocketPrototype.socketRequest = function(method) {
  var i, args = [];
  for (i = 1; i < arguments.length; ++i) {
    args.push(arguments[i]);
  }

  var cb = args.pop();
  args.push(function(body, jwres) {
    successful(jwres) ? cb(null, body, jwres) : cb(makeError(jwres), null, jwres);
  });

  this.io.socket[method.toLowerCase()].apply(this.io.socket, args);
};


///////////////////////////////////////////////////////////////////////////////////////////
/// DQL
///////////////////////////////////////////////////////////////////////////////////////////

/**
 * Select Records
 *
 * @param {String} collectionName
 * @param {Object} options
 * @param {Function} cb
 */
SocketPrototype.select = function(collectionName, options, cb) {

  this.io.socket.get(this.basePath + collectionName, options, function(body, jwres) {
    successful(jwres) ? cb(null, body) : cb(makeError(jwres));
  });

};

/**
 * Insert A Record
 *
 * @param {String} collectionName
 * @param {Object} values
 * @param {Function} cb
 * @return {Object}
 */
SocketPrototype.insert = function(collectionName, values, cb) {

  this.io.socket.post(this.basePath + collectionName, values, function(body, jwres) {
    successful(jwres) ? cb(null, body) : cb(makeError(jwres));
  });

};

/**
 * Update A Record
 *
 * @param {String} collectionName
 * @param {Object} options
 * @param {Object} values
 * @param {Function} cb
 */
SocketPrototype.update = function(collectionName, options, values, cb) {

  var path = this.basePath + collectionName;
  var id = extractId(options, values);

  if (id) {
    // Handling a single update is easy
    this.io.socket.put(path + '/' + id, values, function(body, jwres) {
      successful(jwres) ? cb(null, body) : cb(makeError(jwres));
    });
  }
  else {
    // Handling multiple updates requires the additional custom Blueprint findAndUpdate route
    // See https://github.com/marnusw/sails-hook-semantic-blueprints
    options.data = values;
    this.io.socket.put(path, options, function(body, jwres) {
      successful(jwres) ? cb(null, body) : cb(makeError(jwres));
    });
  }

};

/**
 * Destroy A Record
 *
 * @param {String} collectionName
 * @param {Object} options
 * @param {Function} cb
 */
SocketPrototype.destroy = function(collectionName, options, cb) {

  var path = this.basePath;
  var id = extractId(options);

  if (id) {
    // To handle a single delete send the id to the standard Blueprints route
    path += collectionName + '/' + id;
  }
  else {
    // Deleting multiple records targets the find-n-destroy route and requires
    // the additional custom Blueprint findAndDestroy route.
    // See https://github.com/marnusw/sails-hook-semantic-blueprints
    // The messy URL is necessary since Sails.js already binds `delete /:model`.
    path += 'find-n-destroy' + '/' + collectionName;
  }

  this.io.socket.delete(path, options, function(body, jwres) {
    successful(jwres) ? cb(null, body) : cb(makeError(jwres));
  });

};


///////////////////////////////////////////////////////////////////////////////////////////
/// DDL
///////////////////////////////////////////////////////////////////////////////////////////

/**
 * Describe a collection
 *
 * @param {String} collectionName
 * @param {Function} [cb] If not provided the schema is returned immediately.
 */
SocketPrototype.describe = function(collectionName, cb) {
  var schema = this.collections[collectionName].definition;
  if (!cb) {
    return schema;
  }
  setTimeout(function() {
    cb(null, schema);
  }, 0);
};


///////////////////////////////////////////////////////////////////////////////////////////
/// CONSTRAINTS AND HELPERS
///////////////////////////////////////////////////////////////////////////////////////////

/**
 * Extract the primary key value from search criteria or values.
 *
 * @param  {Object} options
 * @param  {Object|Array} [values] If a single data object is provided it is also checked for an id.
 * @return {Integer|String}
 */
function extractId(options, values) {
  var pk = options.id || (options.where && options.where.id);
  if (!pk && typeof values == 'object') {
    // Even if it's an array or some other instance this check will do no harm.
    pk = values.id;
  }
  // exclude criteria on the id field or multiple ids in an array
  return typeof pk == 'object' ? undefined : pk;
}

/**
 * Inspect the status code to see whether a request was successful or not.
 *
 * @param {Object} jwres
 * @returns {boolean}
 */
function successful(jwres) {
  return jwres.statusCode >= 200 && jwres.statusCode < 300;
}

/**
 * Create status specific error objects based on the status code and response body.
 *
 * @param {Object} jwres
 * @returns {Object}
 */
function makeError(jwres) {
  if (jwres.statusCode == 400) {
    return {
      statusCode: 400,
      type: 'E_VALIDATION',
      validation: true,
      messages: jwres.body
    };
  }

  if (jwres.statusCode == 401) {
    return {
      statusCode: 401,
      type: 'E_UNAUTHORIZED',
      unauthorized: true,
      validation: true,
      messages: jwres.body
    };
  }

  if (jwres.statusCode == 404) {
    return {
      statusCode: 404,
      type: 'E_NOT_FOUND',
      notFound: true,
      messages: jwres.body
    };
  }

  return {
    statusCode: jwres.statusCode,
    type: 'E_FATAL',
    fatal: true,
    message: jwres.body
  };
}
